File: client/js/item.js
define(['entity'], function (Entity) {

  var Item = Entity.extend({
    init: function (id, kind) {
      this._super(id, kind);

      this.wasDropped = false;

      this.amount = 1;

      this._cooldown = 1;
      this._castTime = 1;
      this._tooltip = "N/A";
      this.nameOffsetY += 5;
    },

    get cooldown() {
      return this._cooldown;
    },

    get castTime() {
      return this._castTime;
    },

    get tooltip() {
      return this._tooltip;
    },

    get itemKind() {
      return Types.getKindAsString(this.kind);
    },

    get type() {
      return Types.getType(this.kind);
    },

    get isStackable() {
      return Types.isStackable(this.kind);
    },

    hasShadow: function () {
      return true;
    },

    onLoot: function (player) {
      if (this.type === "weapon") {
        player.lootedWeapon(this);
      } else if (this.type === "armor") {
        player.lootedArmor(this);
      }
    },

    getSpriteName: function () {
      return "item-" + this.itemKind;
    },

    getLootMessage: function () {
      return this.lootMessage;
    }
  });

  return Item;
});

File: client/dart/item.dart
library item;

import "entity.dart";
import "hero.dart";
import "../shared/dart/gametypes.dart";

abstract class Item extends Entity {

  bool wasDropped = false;
  int amount = 1;
  int cooldown = 1;
  int castTime = 1;
  String tooltip = "N/A";
  String lootMessage;
  List<int> playersInvolved = [];

  Item(
    int id, 
    Entities kind, 
    [String name, 
    String this.lootMessage]
  ): super(id, kind) {
    this.name = name;
    this.nameOffsetY += 5;
  }

  String get itemKind => Types.getKindAsString(this.kind);

  String get type => Types.getType(this.kind);

  bool get isStackable => Types.isStackable(this.kind);

  bool hasShadow() => true;

  String getSpriteName() => "item-${this.itemKind}";
  
  void looted(Hero hero);
}

class Weapon extends Item {

  Weapon(int id, Entities kind, String name, String lootMessage): super(id, kind, name, lootMessage);

  void looted(Hero hero) {
    hero.lootedWeapon(this);
  }
}

class Armor extends Item {

  Armor(int id, Entities kind, String name, String lootMessage): super(id, kind, name, lootMessage);

  void looted(Hero hero) {
    hero.lootedArmor(this);
  }
}
